ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplStatsTracker.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.exec.repl;
+
+import org.apache.commons.collections4.map.ListOrderedMap;
+import org.apache.commons.math3.stat.descriptive.DescriptiveStatistics;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Tracks the replication statistics per event type.
+ */
+public class ReplStatsTracker {
+
+  // Maintains the descriptive statistics per event type.
+  private HashMap<String, DescriptiveStatistics> descMap;
+
+  // Maintains the top K costliest eventId's
+  private HashMap<String, ListOrderedMap<Long, Long>> topKEvents;
+  // Number of top events to maintain.
+  private final int k;
+
+  public ReplStatsTracker(int k) {
+    this.k = k;
+    descMap = new HashMap<>();
+    topKEvents = new HashMap<>();
+  }
+
+  /**
+   * Adds an entry for tracking.
+   * @param eventType the type of event.
+   * @param eventId the event id.
+   * @param timeTaken time taken to process the event.
+   */
+  public void addEntry(String eventType, String eventId, long timeTaken) {
+    // Update the entry in the descriptive statistics.
+    DescriptiveStatistics descStatistics = descMap.get(eventType);
+    if (descStatistics == null) {
+      descStatistics = new DescriptiveStatistics();
+      descStatistics.addValue(timeTaken);
+      descMap.put(eventType, descStatistics);
+    } else {
+      descStatistics.addValue(timeTaken);
+    }
+
+    // Tracking for top K events, Maintain the list in descending order.
+    ListOrderedMap<Long, Long> topKEntries = topKEvents.get(eventType);
+    if (topKEntries == null) {
+      topKEntries = new ListOrderedMap<>();
+      topKEntries.put(Long.parseLong(eventId), timeTaken);
+      topKEvents.put(eventType, topKEntries);
+    } else {
+      // Get the index of insertion, by descending order.
+      int index = Collections.binarySearch(new ArrayList(topKEntries.values()), timeTaken, Collections.reverseOrder());
```

Comment:
where are you sorting this?

Reply:
Here itself. :-)
When the first element the list is sorted(only one element), post that this gives the index where to insert so the list stays sorted, so when I insert the element to the index returned from this, the list stays always sorted.